### PR TITLE
Fix/Simplify/Optimize the transformer API for causal and masked pre-training approach

### LINKED ISSUE #918: Support masking for causal language modeling
The current API does not generate the `_keras_mask` for causal language modeling.  This implementation has two constraints: 
- During training and evaluation: The loss is computed for all positions (even the padded index)
- During inference, the model returns the scores for all positions (actual input candidates + padded position). While we usually expect the model to return the score only for one position (the last non-padded one).
- Simplify the masking API of transformer-based models by adding `masking` string argument to the TransformerBlock and hide the custom `pre` blocks such as: `SequenceMaskLastInference` and `ReplaceMaskedEmbeddings`

_Originally posted by @gabrielspmoreira in https://github.com/NVIDIA-Merlin/models/pull/909#pullrequestreview-1214326395_
      
In addition,  for CLM, we should be able to train with `SequencePredictNext` and evaluate with `SequencePredictLast`.

### PR DESCRIPTION
**Pull request by sararb**
Fixes #918 
- It is built on top of the 23.02 release branch to be able to merge it with the main branch of the tutorial image ( it will use the latest release container)

This PR aims to address the following limitations in the current Transformer API:

1. **Inference for CausalLM is not supported**: the model returns scores for all positions in the input sequence but we are only interested on the last position at inference time.
2. **CLM only supports the transform SequencePredictNext**: We cannot train or evaluate a CLM model on the last item of the sequence only (i.e using SequencePredictLast)
3. **Specialized and complex API**: the support of masking and inference is done via specialized Merlin Blocks (`ReplaceMaskedEmbeddings`, `SequenceMaskLastInference`). So the user should be familiar with all these custom blocks to correctly define and train a transformer-based model with CLM or MLM approach.
4. **Output of the model is a padded dense tensor of scores**: As the HuggingFace transformer layer is requiring a dense input. We are converting the ragged inputs to dense (by 0-padding to the maximum length seen in the given batch) right before calling the HF layer. The ops that follow this block are then applied to the padded dense tensor. This means that we compute the logit scores for all positions (even the padded ones) and which can be costly (such in the weight tying multiplication between the hidden representation all items embeddings).

- With this PR, you can train a transformer model with masked language modeling as follows: 
```
    model = mm.Model(
        mm.InputBlockV2(...),
        mm.MLPBlock([64]),
        XLNetBlock(
            d_model=64,
            n_head=8,
            n_layer=2,
            masking="masked",
        ),
        mm.CategoricalOutput(
            seq_schema.select_by_name(target),
            default_loss="categorical_crossentropy",
        ),
    )
    seq_mask_random = mm.SequenceMaskRandom(schema=seq_schema, target=target, masking_prob=0.3)
    model.compile(...)
    model.fit(..., pre=seq_mask_random)

```

* I created this doc for more details about the Merlin TransformerAPI: https://docs.google.com/document/d/1auNJCOPFjyVxlhx4eUUthpMLm-rva0jC-lo0ChSZWzI/edit#

### Goals :soccer:
- [x] Add Inference support to a transformer-based model trained with CLM: i.e. select the prediction score of the last non-padded position.  

- [x] Support evaluating a transformer-based model trained with CLM (trained with SequencePredictNext) on the last item in the sequence (i.e using SequencePredictLast). 

- [x] Simplify the transformer API by adding a string parameter `masking` to indicate whether we train with a `masked` or `causal` approach. 

- [x] Optimize the predictions generation of transformer-based model: Convert the dense tensor returned by the transformer layer to a `tf.RaggedTensor` so that all the logic happening after the transformer block (MLP projections, softmax layer...) is applied only on actual positions.

- [ ] Add checks to ensure the specified `masking` is used with the correct `pre` in fit/evaluate methods.

- [ ] Implement an RNNBlock to support training with SequencePredictNext but evaluate on last position only (SequencePredictLast)

- [ ] Create documentation about session-based models (RNN and Transformers).

- [ ] Update the example notebooks 

### Implementation Details :construction:

- Add a `masking` argument to the TransformerBlock class to abstract the custom blocks needed for training with `causal` of `masked` language modeling. 

- Add masking support for `SequencePredictNext` and `SequencePredictLast`

- Implement the following specialized blocks (needed for CLM and MLM support): 
    *  `SequenceCausalLastInference` to generate a mask indicating the last non-padded position of each input sequence at inference time. 
    * `ExtractMaskFromTargets` to move the logic defined in `ReplaceMaskedEmbeddings` and which consists of inferring the mask information based on targets. 
    * `TransformerOutputToRagged` to convert the output of the transformer layer to Ragged based on the mask information. 

- Added a new method in `_extract_masked_predictions` to extract predictions at target positions when targets are scalars (i.e 1-D).  This is needed to support evaluating a transformer model trained with CLM (i.e. SequencePredictNext) on the last item of the sequence only (i.e SequencePredictLast). 

### Testing Details :mag:
- Update `test_transformer_with_masked_language_modeling` and `test_transformer_with_causal_language_modeling` to account for the changes. 

## Dependency: 
https://github.com/NVIDIA-Merlin/models/pull/983

### CODE REVIEW

File: merlin/models/tf/models/base.py [hunk 1]
```diff
@@ -812,60 +812,172 @@ def call_train_test(
...
+                tf.equal(tf.rank(target), 1),
+                tf.equal(tf.rank(prediction), 3),
+            )
+            prediction = tf.cond(
+                rank_check, lambda: self._extract_masked_predictions(prediction), lambda: prediction
+            )
+
+        # Take the flat values of predictions and targets as Keras
+        # losses does not support RaggedVariantTensor on GPU:
+        prediction = prediction.flat_values
```
Comment (gabrielspmoreira) [inline merlin/models/tf/models/base.py:925]: I think that by using `flat_values` for both `predictions` and `target` we misalign them with the shape of `sample_weight`. In that case, I think we should also flatten `sample_weight` if it has the same rank ?

File: merlin/models/tf/models/base.py [hunk 2]
```diff
@@ -812,60 +812,172 @@ def call_train_test(
...
+            )
+
+        # Ensuring targets and preds have the same dtype
+        target = tf.cast(target, prediction.dtype)
+
+        # If targets are scalars (1-D) and predictions are sequential (3-D),
+        # extract predictions at target position because Keras expects
+        # predictions and targets to have the same shape.
+        if getattr(prediction, "_keras_mask", None) is not None:
+            rank_check = tf.logical_and(
```
Comment (gabrielspmoreira) [inline merlin/models/tf/models/base.py:915]: Do we still need this logic, as Transformer blocks will have the expected blocks to mask the last item and make them 2D?
I mean, couldn't we have the logic inside `_extract_masked_predictions` within `TransformerInferenceHiddenState`, that would work for both inference and evaluation when pre=SequenceMaskLast is used?

File: merlin/models/tf/utils/tf_utils.py [hunk 3]
```diff
@@ -469,3 +469,29 @@ def list_col_to_ragged(col: Tuple[tf.Tensor, tf.Tensor]):
...
+        The input tensor, which can be either a dense or ragged tensor.
+    mask : Union[tf.Tensor, tf.RaggedTensor]
+        The mask tensor, which can be either a dense or ragged tensor.
+
+    Returns
+    -------
+    bool: True if the shape of the input and mask tensors are compatible, False otherwise.
+    """
+    result = False
+    if type(inputs) == type(mask) and (inputs.shape.as_list()[:-1] == mask.shape.as_list()):
```
[the remark this answers is not in the document]
Comment (sararb) [reply merlin/models/tf/utils/tf_utils.py]: Thanks for your comment! The function was originally implemented in `ReplaceMaskedembeddings` to check the 2D mask because it currently only supports 3-D sequential inputs. I agree it would be more useful to make the function more generic to support inputs of different dimensions. 
I updated the code to use `inputs.shape.as_list()[:len(mask.shape)]` instead of inputs.shape.as_list()[:-1] to make the function more flexible.

File: merlin/models/tf/utils/tf_utils.py [hunk 4]
```diff
@@ -469,3 +469,29 @@ def list_col_to_ragged(col: Tuple[tf.Tensor, tf.Tensor]):
         row_lengths = tf.cast(row_lengths, tf.int32)
 
     return tf.RaggedTensor.from_row_lengths(values, row_lengths)
+
+
+def check_inputs_mask_compatible_shape(
```
[the remark this answers is not in the document]
Comment (sararb) [reply merlin/models/tf/utils/tf_utils.py:474]: If the inputs and mask don’t have the same type, the function returns `False`. So we could maybe raise the exception inside the block that uses the util function, rather than modifying the function itself? Like [here](https://github.com/NVIDIA-Merlin/models/blob/95f48f61f5381ff583f9c062b3dad672bd271847/merlin/models/tf/transforms/sequence.py#L812
) in the `ReplaceMaskedembeddings` block.

File: merlin/models/tf/models/base.py [hunk 5]
```diff
@@ -812,60 +812,172 @@ def call_train_test(
...
+        This methods applies the following transformations to the target and prediction tensors:
+        - Select ragged targets based on the mask information, if defined.
+        - Remove mask information from the ragged targets and predictions.
+        - One-hot encode targets if their rank is one less than the rank of predictions.
+        - Ensure that targets have the same shape and dtype as predictions.
+        """
+        if isinstance(target, tf.RaggedTensor) and getattr(target, "_keras_mask", None) is not None:
+            # Select targets at masked positions and return
+            # a ragged tensor.
+            target = tf.ragged.boolean_mask(
```
[the remark this answers is not in the document]
Comment (sararb) [reply merlin/models/tf/models/base.py:910]: This code is designed to work with ragged targets and extract a subset of targets based on masked positions (`_keras_mask`). This is necessary for `SequenceMaskLast` and `SequenceMaskRandom` which copy the whole input sequence of candidates-ids as targets but the model outputs a ragged prediction only for masked positions. So This logic is needed to align the predictions and targets.

File: merlin/models/tf/models/base.py (hunk 1 above)
Comment (sararb) [reply merlin/models/tf/models/base.py:919]: I had a similar [question]( https://github.com/NVIDIA-Merlin/models/pull/981#discussion_r1114558548 ) but couldn’t really figure out where to transform the `sample_weights`. One possible solution could be to add sample_weight as an additional argument to the `adjust_predictions_and_targets` function and apply the same transformations to align it with predictions and targets. What do you think?

**Comment by sararb** [comment]
Closing this PR as the commits history includes other external commits too and it is hard to review the current changes. I will open a new PR with only the changes related to the new design of the transformer API
